### PR TITLE
add training

### PR DESCRIPTION
**Pull request by rrevoid**

### CODE REVIEW

File: scripts/train_model.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    parser.add_argument("-m", "--diffusion", help="Base diffusion name from huggingface")
+    parser.add_argument("-p", "--path", help="Path to dataset for training")
+    parser.add_argument("-bs", "--batch_size", type=int, default=64, help="Batch size for training")
+    parser.add_argument("-s", "--side", default="left", help="Side of truncation and padding for tokenizer")
+    parser.add_argument("-cl", "--max_context", type=int, default=256, help="Max length of context fed into diffusion")
+    parser.add_argument("-gl", "--max_gen", type=int, default=64, help="Max length of generated output")
+    parser.add_argument("-d", "--device", default="cpu", help="Device on which to evaluate the diffusion")
+    parser.add_argument("-sc", "--schedule", default="linear", help="Noise schedule for diffusion diffusion")
+    parser.add_argument("-df", "--diffusion_steps", type=int, default=5, help="Number of diffusion steps")
+    parser.add_argument("-e", "--num_epochs", type=int, default=1, help="Number of training epochs")
```
Comment (SpirinEgor) [inline scripts/train_model.py]: It seems that not all arguments are required for training or have poor naming:
- `--diffusion` -> `model` or `base_model`. Also, fix the help message.
- `-p`/`--path` -> `dataset` or `dataset_dir`
- The side of truncation has the default value inside the tokenizer, no need to specify it.
- `--max_gen` let's call it `decoder_context_size`. Since the output is always this size.
- `--num_epochs` let's measure training in steps, it is much more convenient for large models and lets you precisely calculate a number of processed tokens.

Comment (SpirinEgor) [reply scripts/train_model.py]: By the way, what do you think about replacing a large arg parser with something like YAML? You can parse it with [`OmegaConf`](https://omegaconf.readthedocs.io/en/2.3_branch/).

You can version config as well as code. And it is much easier to run training:
```shell
python -m scripts.train configs/train.yaml $DATASET_DIR
```

File: src/diffusion/model.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+class DiffusionTransformer(nn.Module):
+    def __init__(self, name: str, diffusion_steps: int):
+
+        super().__init__()
+
+        model = AutoModel.from_pretrained(name)
+
+        self.emb = self._get_embeddings(model)
+        self.time_embeds = self._get_time_embeddings(model, diffusion_steps)
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: No need for a separate method, it makes the class more complex. Try to keep it simple.
```python
self.time_embeds = nn.Embedding(diffusion_steps, self.emb.embedding_dim)
```

File: src/diffusion/model.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        self.encoder = self._get_encoder(model)
+        self.decoder = model.decoder
+
+    @staticmethod
+    def _freeze_params(model):
+        for parameter in model.parameters():
+            parameter.requires_grad = False
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: Just move this method out of function. This is something general for PyTorch, not class-specific.

File: src/diffusion/model.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        return nn.Embedding(input_dim, emb_dim)
+
+    def _get_encoder(self, model):
+        encoder = model.encoder
+        self._freeze_params(encoder)
+
+        return encoder
+
+    def get_embeddings(self, input_ids):
+        return self.emb(input_ids)
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: Why do you need this function?

File: src/diffusion/model.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        encoder_attention_mask=None,
+        decoder_inputs_embeds=None,
+        time_ids=None,
+    ):
+        context = self.encoder(input_ids=encoder_input_ids, attention_mask=encoder_attention_mask).last_hidden_state
+
+        time_embeds = self.time_embeds(time_ids)
+        input_embeds = decoder_inputs_embeds + time_embeds
+
+        decoder_attention_mask = torch.ones(input_embeds.shape[:2])
```
Comment (SpirinEgor) [inline src/diffusion/model.py]: You can omit it, HF will by default use this mask

File: src/diffusion/train_utils.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+def get_diffusion_variables(
+    model,
+    gt,
+    max_gen: int,
+    diffusion_steps: int,
+    alphas_cumprod_prev: torch.Tensor,
+    device: str,
+):
+    pad_emb = model.get_embeddings(torch.LongTensor([0]))
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: - Pad may have another index. I think you specify it by checking the id in the tokenizer.
- I suggest moving this function inside `DiffusionTransformer` (btw, `Seq2SeqDiffusionTransformer` would suit better). You can save the pad vector to the module state (self._pad_emb = ...) and reuse it.

File: src/diffusion/train_utils.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+    x_0 = model.get_embeddings(gt.input_ids)
+    padding = pad_emb.repeat((x_0.shape[0], max_gen - x_0.shape[1], 1))
+    x_0 = torch.concat((x_0, padding), dim=1)
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: Will this work for the batch of `x_0`?

File: src/diffusion/train_utils.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+    t = torch.randint(diffusion_steps, size=(1,))
+
+    x_t = torch.sqrt(alphas_cumprod_prev[t]) * x_0 + torch.sqrt(1 - alphas_cumprod_prev[t]) * torch.normal(
+        0, 1, size=x_0.shape
+    )
+
+    return x_0.to(device), x_t.to(device), t.to(device)
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: Let's move to a device outside the function. This will simplify interface.

**Review by SpirinEgor — CHANGES_REQUESTED** [review]
This is quite a good code for training the model 👍. Let's prettify it a little bit.

As general feedback, we definitely need to integrate `wandb` to log training info (e.g. to see how loss decrease). Also, at some point, we will need to add an evaluation step to measure metrics.

File: src/diffusion/train_utils.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    schedule: str,
+    diffusion_steps: int,
+    num_epochs: int,
+    max_context: int,
+    max_gen: int,
+    device: str,
+    project_name: str = "didi",
+    run_name: str = "Training diffusion",
+):
+    wandb.init(project=project_name, name=run_name)
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: Let's keep the project name random. But you can specify the group name. It is useful for training models with different sizes, for example.

`Training diffusion` -- all your runs are about training diffusion, so this name does not make sense.

File: src/diffusion/train_utils.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+    optimizer = torch.optim.Adam(model.parameters())
+
+    alphas_cumprod_prev = configure_schedule(diffusion_steps, schedule)
+
+    for _ in tqdm(range(num_epochs), desc="Training"):
+        for batch in tqdm(dataloader, desc="Epoch"):
+            context = tokenizer(
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: But you already tokenize data inside collate function 🧐

File: src/diffusion/train_utils.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+                encoder_attention_mask=context.attention_mask,
+                decoder_inputs_embeds=x_t,
+                time_ids=t,
+            )
+
+            optimizer.zero_grad()
+
+            loss = F.mse_loss(x_0_hat, x_0)
+
+            wandb.log({"loss": loss.item()})
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: Logging information on each step can load the network. It's better to keep a log with some periodicity, e.g. every 10 steps.

File: src/diffusion/train_utils.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    alphas_cumprod_prev = F.pad(alphas_cumprod, (1, 0), value=1.0)
+    return alphas_cumprod_prev
+
+
+def get_xt(x_0, alphas_cumprod_prev, t, device):
+    alphas_cumprod_prev_t = alphas_cumprod_prev[t].reshape(-1, 1, 1)
+
+    x_t = torch.sqrt(alphas_cumprod_prev_t) * x_0 + torch.sqrt(1 - alphas_cumprod_prev_t) * torch.normal(
+        0, 1, size=x_0.shape
+    ).to(device)
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]:
```python
noise = torch.normal(0, 1, size=x_0.shape).to(device)
x_t = torch.sqrt(alphas_cumprod_prev_t) * x_0 + torch.sqrt(1 - alphas_cumprod_prev_t) * noise
```

File: src/diffusion/train_utils.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+    return x_t
+
+
+def prepare_x0(model, emb, max_gen, device):
+    sigma_0 = 0.1
+
+    pad_emb = model.emb(torch.tensor(model.emb.padding_idx).to(device))
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: `x_0` is already padded and have shape `[batch size; target seq len]`. So no need for additional padding here.

File: src/diffusion/train_utils.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    padding = pad_emb.repeat((x_0.shape[0], max_gen - x_0.shape[1], 1)).to(device)
+    x_0 = torch.concat((x_0, padding), dim=1)
+
+    x_0 += torch.normal(0, sigma_0, size=x_0.shape).to(device)
+
+    return x_0
+
+
+def get_diffusion_variables(
+    model,
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: Pass diffusion steps here directly, this simplify function signature

File: src/diffusion/train_utils.py [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+    x_0 = prepare_x0(model, emb, max_gen, device)
+
+    t = torch.randint(1, model.diffusion_steps + 1, size=(x_0.shape[0],)).to(device)
+
+    x_t = get_xt(x_0, alphas_cumprod_prev, t, device)
+
+    return x_0, x_t, t
+
+
+def collate_gt(list_of_samples, tokenizer, max_context, max_gt):
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: The dataset class already has a collator, so no need for this function

File: src/diffusion/train_utils.py [hunk 16]
```diff
@@ -0,0 +1,152 @@
+import torch
+import torch.nn.functional as F
+import wandb
+from tqdm import tqdm
+
+from schedules import cosine_beta_schedule
+from schedules import linear_beta_schedule
+from metrics import calculate_ce
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]:
```python
from src.diffusion.schedules import cosine_beta_schedule, linear_beta_schedule
from src.metrics import calculate_ce
```

File: src/diffusion/train_utils.py [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+    wandb.init(project=project_name)
+
+    optimizer = torch.optim.Adam(model.parameters())
+
+    alphas_cumprod_prev = configure_schedule(model.diffusion_steps, schedule).to(device)
+
+    model.train()
+    with tqdm(total=num_steps) as pbar:
+        while num_steps:
+            for b_context, b_gt in tqdm(train_dataloader):
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: Shapes:
- `b_context`: `[batch size; context seq len]`
- `b_gt`: `[batch size; target seq len]`

They already padded

File: src/diffusion/train_utils.py [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+                probs = model.classifier(x_0)
+
+                target = F.pad(
+                    gt.input_ids, (0, probs.shape[1] - gt.input_ids.shape[1]), "constant", model.emb.padding_idx
+                )
+                pad_mask = target == model.emb.padding_idx
+
+                target[pad_mask] = -100
+                ce = F.cross_entropy(torch.transpose(probs, 1, 2), target)
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: This is a code duplicate (PyCharm highlighted this), you should move it to a separate function

File: src/diffusion/train_utils.py [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+                    time_ids=t,
+                )
+
+                optimizer.zero_grad()
+
+                probs = model.classifier(x_0)
+
+                target = F.pad(
+                    gt.input_ids, (0, probs.shape[1] - gt.input_ids.shape[1]), "constant", model.emb.padding_idx
+                )
```
Comment (SpirinEgor) [inline src/diffusion/train_utils.py]: No need in padding, dataset already handle it

File: src/metrics.py [hunk 20]
```diff
@@ -3,6 +3,9 @@
 from loguru import logger
 from tqdm.auto import tqdm
 
+from train_utils import get_xt
+from train_utils import prepare_x0
```
Comment (SpirinEgor) [inline src/metrics.py]: And you import smth from `metrics` in `train_utils`, so you would get circular import error

File: src/data/dataset.py [hunk 21]
```diff
@@ -17,15 +18,21 @@ class ConvAI2Dataset(Dataset):
...
 
-    def __init__(self, path, tokenizer, max_seq_len):
+    def __init__(self, path, tokenizer_name, max_context_len, max_target_len=None, have_candidates=True):
         self.dataset = []
         self.num_dialogs = 0
 
-        bos = tokenizer.bos_token
-        eos = tokenizer.eos_token
+        self.context_tokenizer = AutoTokenizer.from_pretrained(tokenizer_name, truncation_side="left")
+        self.candidate_tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
```
Comment (SpirinEgor) [inline src/data/dataset.py:26]: Probably, you can specify max length here. And let's set truncation side for candidate tokenizer explicitly.

File: src/diffusion/model.py [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+
+        self.emb = nn.Embedding(vocabulary_size, emb_dim, padding_idx=0)
+        self.time_embeds = nn.Embedding(diffusion_steps + 1, emb_dim)
+
+        self.encoder = model.encoder
+        self.decoder = model.decoder
+
+        self.classifier = nn.Linear(emb_dim, vocabulary_size)
+
+        freeze_params(self.encoder)
```
Comment (SpirinEgor) [inline src/diffusion/model.py:30]: I guess we also need to set `self.encoder.eval()`

File: src/metrics.py [hunk 23]
```diff
@@ -86,3 +89,54 @@ def calculate_f1(model, tokenizer, dataloader, max_gen_len, do_sample, num_beams
...
+                time_ids=t,
+            )
+
+        x_1 = get_xt(x_0, alphas_cumprod_prev, ones, device)
+
+        x_0 = model(
+            encoder_input_ids=context,
+            decoder_inputs_embeds=x_1,
+            time_ids=ones,
+        )
```
Comment (SpirinEgor) [inline src/metrics.py:136]: This is the same as the last iteration of the loop. Why are you repeating this?

File: src/training.py [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+                x_0_hat = model(
+                    encoder_input_ids=context,
+                    decoder_inputs_embeds=x_t,
+                    time_ids=t,
+                )
+
+                optimizer.zero_grad()
+
+                pad_mask = gt == model.emb.padding_idx
+                ce = calculate_batch_ce(model, x_0, gt, pad_mask)
```
Comment (SpirinEgor) [inline src/training.py:51]: Are you sure that we should calculate loss using `x_0`? I looked into the Diffusion-LM code, the use model output for this (`x_0_hat` in our case) -- [code](https://github.com/XiangLi1999/Diffusion-LM/blob/main/improved-diffusion/improved_diffusion/gaussian_diffusion.py#L1688)

Comment (SpirinEgor) [reply src/training.py:51]: Hmm, but other implementation uses `x_0` -- [code](https://github.com/maximzubkov/minimal-text-diffusion/blob/main/src/modeling/diffusion/gaussian_diffusion.py#L295)

File: src/training.py [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+
+                emb_mask = ~pad_mask.unsqueeze(-1)
+
+                mse = torch.where(
+                    t == 1,
```
Comment (SpirinEgor) [inline src/training.py:56]: `t == 0`?

**Review by SpirinEgor — APPROVED** [review]
👍
